File: Whatsapp/OpenWA/sdk/python/openwa/resources/channels.py
"""Channels resource — WhatsApp Channels / Newsletters.

Backed by ``src/modules/channel/channel.controller.ts``
(``@Controller('sessions/:sessionId/channels')``).
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from .._http import quote_segment
from ..types import (
    ChannelMessageQuery,
    ChannelMessageRecord,
    ChannelRecord,
    SubscribeChannelRequest,
    SuccessResult,
)

if TYPE_CHECKING:
    from .._http import HttpExecutor


class ChannelsResource:
    def __init__(self, http: "HttpExecutor") -> None:
        self._http = http

    def list(self, session_id: str) -> list[ChannelRecord]:
        return self._http.request("GET", f"/api/sessions/{quote_segment(session_id)}/channels")

    def get(self, session_id: str, channel_id: str) -> ChannelRecord:
        return self._http.request("GET", f"/api/sessions/{quote_segment(session_id)}/channels/{quote_segment(channel_id)}")

    def messages(
        self, session_id: str, channel_id: str, query: ChannelMessageQuery | None = None
    ) -> list[ChannelMessageRecord]:
        return self._http.request(
            "GET", f"/api/sessions/{quote_segment(session_id)}/channels/{quote_segment(channel_id)}/messages", query=query
        )

    def subscribe(self, session_id: str, body: SubscribeChannelRequest) -> ChannelRecord:
        """Subscribe to a channel using its invite code. Requires an OPERATOR-level key."""
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/channels/subscribe", body=body)

    def unsubscribe(self, session_id: str, channel_id: str) -> SuccessResult:
        """Unsubscribe from a channel. Requires an OPERATOR-level key."""
        return self._http.request("DELETE", f"/api/sessions/{quote_segment(session_id)}/channels/{quote_segment(channel_id)}")

File: Whatsapp/OpenWA/sdk/javascript/src/resources/channels.ts
/**
 * Channels resource — WhatsApp Channels / Newsletters.
 *
 * Backed by `src/modules/channel/channel.controller.ts` (`@Controller('sessions/:sessionId/channels')`).
 * @packageDocumentation
 */

import { encodeSegment } from '../http.js';
import type { OpenWAClient } from '../client.js';
import type {
  ChannelMessageQuery,
  ChannelMessageRecord,
  ChannelRecord,
  SubscribeChannelRequest,
  SuccessResult,
} from '../types.js';

export class ChannelsResource {
  constructor(private readonly client: OpenWAClient) {}

  /** List all channels/newsletters the session is subscribed to. */
  list(sessionId: string): Promise<ChannelRecord[]> {
    return this.client.request<ChannelRecord[]>({
      method: 'GET',
      path: `/api/sessions/${encodeSegment(sessionId)}/channels`,
    });
  }

  /** Get a single channel by id. */
  get(sessionId: string, channelId: string): Promise<ChannelRecord> {
    return this.client.request<ChannelRecord>({
      method: 'GET',
      path: `/api/sessions/${encodeSegment(sessionId)}/channels/${encodeSegment(channelId)}`,
    });
  }

  /** Get recent messages from a channel. */
  messages(sessionId: string, channelId: string, query?: ChannelMessageQuery): Promise<ChannelMessageRecord[]> {
    return this.client.request<ChannelMessageRecord[]>({
      method: 'GET',
      path: `/api/sessions/${encodeSegment(sessionId)}/channels/${encodeSegment(channelId)}/messages`,
      query,
    });
  }

  /** Subscribe to a channel using its invite code. Requires an OPERATOR-level key. */
  subscribe(sessionId: string, body: SubscribeChannelRequest): Promise<ChannelRecord> {
    return this.client.request<ChannelRecord>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/channels/subscribe`,
      body,
    });
  }

  /** Unsubscribe from a channel. Requires an OPERATOR-level key. */
  unsubscribe(sessionId: string, channelId: string): Promise<SuccessResult> {
    return this.client.request<SuccessResult>({
      method: 'DELETE',
      path: `/api/sessions/${encodeSegment(sessionId)}/channels/${encodeSegment(channelId)}`,
    });
  }
}
